fish-rust/src/common.rs

```diff
@@ -1642,55 +1649,61 @@
     // overhead since there's no actual race condition here - even if multiple threads call this
     // routine simultaneously the first time around, we just end up needlessly querying uname(2) one
     // more time.
-    *IS_WINDOWS_SUBSYSTEM_FOR_LINUX
-}
-
-fn slice_contains_slice<T: Eq>(a: &[T], b: &[T]) -> bool {
-    a.windows(b.len()).any(|aw| aw == b)
-}
-
-#[cfg(not(windows))]
-static IS_WINDOWS_SUBSYSTEM_FOR_LINUX: Lazy<bool> = Lazy::new(|| false);
-#[cfg(windows)]
-static IS_WINDOWS_SUBSYSTEM_FOR_LINUX: Lazy<bool> = Lazy::new(|| {
-    let mut info: libc::utsname = unsafe { mem::zeroed() };
-    unsafe {
-        libc::uname(&mut info);
+    #[cfg(not(target_os = "linux"))]
+    {
+        false
     }
 
-    // Sample utsname.release under WSL, testing for something like `4.4.0-17763-Microsoft`
-    if !slice_contains_slice(&info.release, b"Microsoft") {
-        return false;
-    }
-    let dash = info.release.iter().position('-');
+    #[cfg(target_os = "linux")]
+    IS_WINDOWS_SUBSYSTEM_FOR_LINUX.get_or_init(|| {
+        let mut info: libc::utsname = unsafe { mem::zeroed() };
```

Comment:
BTW I always use `mem::zeroed()` to initialize C structs; there ought to be a better way, that hides the `unsafe`..

Reply:
We could make a safe version of zeroed that asserts that the type does not have a destructor nor does it contain any types that have a destructor (recursively), via https://doc.rust-lang.org/std/mem/fn.needs_drop.html

I'm not absolutely certain that means a type is safe to initialize via `mem::zeroed()` but I think it'll be close enough to it for our purposes?

Reply:
great idea, I'll try that. It's not safe for NonZeroInt but we only want to use it for C structs

Reply:
Ugh, it would need to be a macro because you can't use the outer generic type to make a const call, and the type would need to be specified explicitly whereas all our current `zeroed()` uses infer the type. Maybe not worth it.

Reply:
There is no generic way to check if a type can be zero-initialized; if there was, `bytemuck` wouldn't have to manually implement its [`Zeroable`](https://docs.rs/bytemuck/latest/bytemuck/trait.Zeroable.html) trait for all zero-initializable std types.

Reply:
Of course, this was merely an attempt to see if there was some sort of good-enough compromise.

tbh, I don't see a problem with `unsafe { std::mem::zeroed() }` but that was going to be my other suggestion here as well: we can add our own trait and implement it for specific libc types as needed if we really needed to cut down on that unsafe for whatever reason.